include/remill/BC/PCodeCFG.h

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2022-present Trail of Bits, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <glog/logging.h>
+
+#include <cstddef>
+#include <mutex>
+#include <sleigh/libsleigh.hh>
+#include <unordered_map>
+#include <variant>
+#include <vector>
+
+#include "lib/Arch/Sleigh/ControlFlowStructuring.h"
+
+
+namespace remill {
+namespace sleigh {
+
+
+struct IntrainstructionIndex {
+  size_t target_block_ind;
+};
+
+struct InstrExit {};
+
+using Exit = std::variant<IntrainstructionIndex, InstrExit>;
+
+struct ConditionalExit {
+  Exit true_branch;
+  Exit false_branch;
+};
+
+class PcodeBlock {
+ public:
+  size_t base_index;
+  std::vector<RemillPcodeOp> ops;
+  std::variant<Exit, ConditionalExit> block_exit;
+
+  // Default block with an instruction exit and no ops.
+  PcodeBlock(size_t base_index);
+
+  PcodeBlock(size_t base_index, std::vector<RemillPcodeOp> ops,
+             std::variant<Exit, ConditionalExit> block_exit);
+};
+
+class PcodeCFG {
+ public:
+  std::unordered_map<size_t, PcodeBlock> blocks;
+
+  PcodeCFG(std::unordered_map<size_t, PcodeBlock> blocks);
+};
+
+class PcodeCFGBuilder {
+ private:
+  const std::vector<RemillPcodeOp> &linear_ops;
+
+
+  PcodeCFGBuilder(const std::vector<RemillPcodeOp> &linear_ops);
+
+
+  PcodeBlock BuildBlock(size_t start_ind, size_t next_start);
+  std::variant<Exit, ConditionalExit> GetBlockExitsForIndex(size_t index);
+
+  std::vector<size_t> GetIntraProcTargets(size_t index);
+
+  std::vector<size_t> GetBlockStarts();
+
+ public:
+  static PcodeCFG CreateCFG(const std::vector<RemillPcodeOp> &linear_ops);
+
+  PcodeCFG Build();
```

Comment:
This is really confusing. What does `Build` do and why is `CreateCFG` not a ctor? From the type of `Build` I would expect it creates a new instance from current one, but it is used in `CreateCFG` instead.

I would possibly rename `Build` and add documentation.

Reply:
[062a7e2](https://github.com/lifting-bits/remill/pull/645/commits/062a7e2ecc9d56301524f3639f0c6bdb9f8e4c46)

I think the intension should be clearer now. This class just exposes a constructor and a `Build` function now that returns a CFG (and since it's called a CFG builder, that makes sense).

We then have a `CreateCFG` helper which is short hand for constructing a `PcodeCFGBuilder` and immediately calling `Build` on it. I've moved this out of the class since it's just using the public API and there's no need for it to have access to private members, etc.